dockerfiles/Dockerfile

```diff
@@ -65,42 +65,81 @@ RUN apt-get update \
     && DEBIAN_FRONTEND=noninteractive apt-get upgrade -y \
     && DEBIAN_FRONTEND=noninteractive apt-get install -y \
         ca-certificates \
+        curl \
         tini \
     && rm -rf /var/lib/apt/lists/*
 
+WORKDIR /srv/docsrs
+
+# Tini is a small init binary to properly handle signals
+CMD ["/usr/bin/tini", "/usr/local/bin/cratesfyi", "start-web-server", "0.0.0.0:80"]
+
 COPY --from=build /build/target/release/cratesfyi /usr/local/bin
 COPY static /srv/docsrs/static
 COPY templates /srv/docsrs/templates
 COPY vendor /srv/docsrs/vendor
 
-WORKDIR /srv/docsrs
+########################
+#  Build server stage  #
+########################
+
+FROM ubuntu:22.04 AS build-server
+
+RUN apt-get update \
+    && DEBIAN_FRONTEND=noninteractive apt-get upgrade -y \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y \
+        ca-certificates \
+        tini \
+        curl \
+        docker.io \
+        build-essential \
+        gcc \
+        pkg-config \
+        libssl-dev \
+    && rm -rf /var/lib/apt/lists/*
+
 # Tini is a small init binary to properly handle signals
-CMD ["/usr/bin/tini", "/usr/local/bin/cratesfyi", "start-web-server", "0.0.0.0:80"]
+CMD ["/usr/bin/tini", "/usr/local/bin/cratesfyi", "start-build-server"]
+
+COPY --from=build /build/target/release/cratesfyi /usr/local/bin
 
-##################
-#  Output stage  #
-##################
+############################
+#  Registry watcher stage  #
+############################
 
-FROM ubuntu:22.04 AS output
+FROM ubuntu:22.04 AS registry-watcher
 
-RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
-    git \
-    libmagic1 \
-    docker.io \
-    ca-certificates \
-    build-essential \
-    gcc \
-    pkg-config \
-    libssl-dev
+RUN apt-get update \
+    && DEBIAN_FRONTEND=noninteractive apt-get upgrade -y \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y \
+        ca-certificates \
+        tini \
+        curl \
+        git \
+    && rm -rf /var/lib/apt/lists/*
+
+# Tini is a small init binary to properly handle signals
+CMD ["/usr/bin/tini", "/usr/local/bin/cratesfyi", "start-registry-watcher", "--repository-stats-updater=enabled", "--cdn-invalidator=enabled"]
+
+COPY --from=build /build/target/release/cratesfyi /usr/local/bin
+
+###############
+#  CLI stage  #
+###############
+
+FROM ubuntu:22.04 AS cli
+
+RUN apt-get update \
+    && DEBIAN_FRONTEND=noninteractive apt-get upgrade -y \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y \
+        ca-certificates \
+        docker.io \
```

Comment:
related to the other question above, I wonder if we could limit the CLI stage to just access the same things as the webserver. 

( with the goal in mind that every stage has its own binary with only the dependencies it needs)